### PR TITLE
Allow admins to get all visible resumes

### PR DESCRIPTION
**Pull request by dowhep**
Create API routes for getting all opted in/visible resumes

### CODE REVIEW

File: repositories/ResumeRepository.ts
Comment (shanekim28) [inline repositories/ResumeRepository.ts:1]: I've found that VSCode's auto-importer fails when using absolute routes. Make sure you use the relative route like `../models/ResumeModel` whenever you're importing.

File: models/UserModel.ts
Comment (shanekim28) [inline models/UserModel.ts:137]: Do we need this? Since all resumes should be handled through the controller, I feel like we don't need a field for this in the UserModel. The "Resumes" table and controller should handle all of the Resume viewing/updating for a single user

File: api/controllers/ResumeController.ts
Comment (shravanhariharan2) [inline api/controllers/ResumeController.ts:24]: I think we should have some kind of permissions check here. e.g., only admins should be able to get all visible resumes, not any member.

File: repositories/ResumeRepository.ts
Comment (shravanhariharan2) [inline repositories/ResumeRepository.ts:10]: nit: let's name it something like `findVisibleResumes()`

File: repositories/ResumeRepository.ts
Comment (shravanhariharan2) [inline repositories/ResumeRepository.ts:21]: nit: let's rename to `findAllByUser` since this would return an array as opposed to a single object

File: services/ResumeService.ts
Comment (shravanhariharan2) [inline services/ResumeService.ts:20]: nit: formatting - line 19 should be tabbed inwards

Comment (shravanhariharan2) [reply services/ResumeService.ts:20]: also, do we need to throw an error if no resumes were found? I imagine just returning an empty array should suffice, and thus no resumes would be displayed on the frontend.

File: tests/controllers/ControllerFactory.ts
Comment (shravanhariharan2) [inline tests/controllers/ControllerFactory.ts:19]: nit: this should be a relative import

File: models/UserModel.ts
Comment (shravanhariharan2) [reply models/UserModel.ts:137]: I think this is fine - getting a user's own profile should return all the info regarding that user, which includes their resume(s). Resumes are tightly related to a user's profile (since that's one of the only places they're displayed), so I don't particularly see an issue with having them returned in the same API call.

Also, @dowhep I don't believe this code works here because we aren't [eager-loading](https://typeorm.io/eager-and-lazy-relations) resumes. This means that this field will be undefined every time we make this function call. Instead, we need to query the Resumes table for all resumes for that user, and attach it to this field. Then, we can return the fields as you specified.

**Review by shravanhariharan2 — CHANGES_REQUESTED** [review]
Nice work - let's add some quick tests for asserting that getting all visible resumes returns only visible resumes, and that getting a user's full profile returns the resumes properly.

File: services/ResumeService.ts
Comment (dowhep) [reply services/ResumeService.ts:20]: So far I think an empty array should suffice

File: services/ResumeService.ts
Comment (michl1001) [inline services/ResumeService.ts:7]: nit: is there a reason we need this import?

File: tests/resume.test.ts
Comment (michl1001) [inline tests/resume.test.ts:4]: Might be worth handling other common error cases? (request a user resume that isn't uploaded yet, what happens with reuploads)

File: repositories/ResumeRepository.ts
Comment (michl1001) [inline repositories/ResumeRepository.ts:11]: Should we filter duplicates from one person?

**Review by michl1001 — COMMENTED** [review]
Looks good for general logic - though had a few nitpick and questions

File: services/ResumeService.ts
Comment (shravanhariharan2) [reply services/ResumeService.ts:7]: Yeah let's definitely remove any commented-out code

File: api/controllers/ResumeController.ts
Comment (shravanhariharan2) [inline api/controllers/ResumeController.ts:25]: We should throw if someone does not have permission to see all resumes (and if someone doesn't have permission to view stuff in general). If someone who doesn't have permission is calling this API route, this is fundamentally an error by the caller, and is something that we should throw a 401 Forbidden Error for. Rerturning an empty response makes it seem like they _do_ have permission, but that there just aren't any resumes, which is incorrect from a data point of view.

File: repositories/ResumeRepository.ts
Comment (shravanhariharan2) [reply repositories/ResumeRepository.ts:11]: At some point we may support users having multiple resumes, and we also have the guarantee that there'll only be one resume per-person right now due to our resume deletion logic, so I think it's okay to not do any duplicate filtering here (also to keep API types consistent so that we wouldn't have to switch from returning Resume to returning Resume[] if we do need to support multiple resumes.

File: services/ResumeService.ts
Comment (shravanhariharan2) [inline services/ResumeService.ts:19]: Related to my first comment, let's throw on permission error instead of returning an empty response

File: tests/resume.test.ts
Comment (shravanhariharan2) [inline tests/resume.test.ts:53]: Was there a reason this is commented out?

File: models/UserModel.ts
Comment (shravanhariharan2) [inline models/UserModel.ts:126]: Here's my idea: let's not perform the resume lookup in `getFullUserProfile()`, and instead let's create a new function `async getResumes()` that does the lookup and populates the `user.resumes` field. Then, in the getCurrentUser() route, we can call this function and then return `user.getFullUserProfile()` in the API response.

The logic behind this is that there's only one route that would ever *need* to return resumes, which is the getCurrentUser() route. All the other calls to getFullUserProfile() don't actually need the resumes to be populated, which is why I'm hesitant to add an extra query to that function.

Let's still return `resumes: user.resumes` in the getFullUserProfile() return object so that the field is not undefined, but I don't necessarily think it needs to be populated, and if the caller needs to populate it, they can call `getResumes()` before it. Let's also leave a comment on the `getFullUserProfile()` function mentioning this.

Feel free to follow up if there's any questions with the approach

File: models/UserModel.ts
Comment (shravanhariharan2) [inline models/UserModel.ts:140]: nit: this would become `resumes: this.resumes`

**Review by shravanhariharan2 — COMMENTED** [review]
Good progress, let me know if there's questions regarding my or Michael's comments 👍🏾

File: services/ResumeService.ts
Comment (dowhep) [reply services/ResumeService.ts:7]: Was using it at some point and thought i needed it later, gonna remove it now

File: tests/resume.test.ts
Comment (dowhep) [reply tests/resume.test.ts:4]: good idea, will look into that

File: api/controllers/ResumeController.ts
Comment (dowhep) [reply api/controllers/ResumeController.ts:25]: Would you recommend handling the permission error in the controller in this case?

File: tests/resume.test.ts
Comment (dowhep) [reply tests/resume.test.ts:53]: It was for reference for test cases, but I wanted to make sure the resume fetching works in getFullUserProfile before I started with the test cases

File: api/controllers/ResumeController.ts
Comment (shravanhariharan2) [reply api/controllers/ResumeController.ts:25]: Yes, since permissions are related to whether or not we should process a request, permission logic should happen in the controller

File: models/UserModel.ts
Comment (dowhep) [reply models/UserModel.ts]: Is there a case where calling `getFullUserProfile` actually does not want to have the resume field populated even though it was populated with the new `getResume` function?

File: api/controllers/ResumeController.ts
Comment (shravanhariharan2) [inline api/controllers/ResumeController.ts:12]: nit: let's call this `GetVisibleResumesResponse` to match the API name

File: models/ResumeModel.ts
Comment (shravanhariharan2) [inline models/ResumeModel.ts:17]: I think this default should be false, since the client should actually specify if it wants to share the resume publicly before doing so. That way, users wouldn't have their resumes publicly visible unless they specify (which they'd do during the upload of their resume, so it's not actually too much work on their end even if we set the default to false)

File: services/UserAccountService.ts
Comment (shravanhariharan2) [inline services/UserAccountService.ts:155]: Lovely

File: api/controllers/UserController.ts
Comment (shravanhariharan2) [inline api/controllers/UserController.ts:35]: nit: I think resumeService is unused in this controller, so let's remove it

File: tests/resume.test.ts
Comment (shravanhariharan2) [inline tests/resume.test.ts:40]: This resume upload operation can be moved into PortalState. The reason we have PortalState is to make it easy to persist fake data into the database.

Here, we are creating mocks, controllers, etc. to upload a resume but all we care about is that a resume is in the database - we've already validated that the 'upload resume' controller works as intended. So instead, let's modify PortalState to have a `createResumes` function that takes in a user and a list of resumes, and uploads resumes for that user. Note that we'd also need a `ResumeFactory` to create fake resumes.

The corresponding code should look something like this:
```ts
    const conn = await DatabaseConnection.get();
    const admin = UserFactory.fake({ accessType: UserAccessType.ADMIN });
    const member = UserFactory.fake();
    const resume = ResumeFactory.fake({ userId: member.id });

    await new PortalState()
      .createUsers(admin, member)
      .createResumes(member, resume)
      .write();
    
    // throws permissions error for member
    expect(resumeController.getAllVisibleResumes(member))
      .rejects.toThrow(ForbiddenError);

    // get response resumes
    const response = await resumeController.getAllVisibleResumes(admin);
    expect(response.error).toBeNull();
    expect(response.resumes).toContainEqual(uploadedResume);
```

Let me know if there's any questions with what I mentioned!

**Review by shravanhariharan2 — COMMENTED** [review]
Almost there! A few comments mostly pertaining to the tests, implementation looks great.

Also, can we delete the package-lock.json? Looks like `npm` was ran somewhere and the lockfile was committed

File: api/controllers/ResumeController.ts
Comment (dowhep) [reply api/controllers/ResumeController.ts:12]: edited

File: models/ResumeModel.ts
Comment (shravanhariharan2) [reply models/ResumeModel.ts:17]: Also, this is not a change in your PR, but in `ResumeModel::getPublicResume()` we need to return the entire user, and not just their uuid. You can see in the figma below that we need to return some user data to the frontend (in this case, just first and last names):
image

Can we add this change into this PR? Instead of returning just `user: this.user.uuid`, let's return the user as a `PublicUser`. This means we also need to do a join when querying all visible resumes. Let's also test that we're returning the proper user object in our tests as well (just making sure the object is returned instead of just a uuid)

File: tests/resume.test.ts
Comment (shravanhariharan2) [inline tests/resume.test.ts:52]: Instead of calling `delete` on objects to strip down the fields we don't want to test, let's instead make it clear the fields we do want to test. This follows the testing convention we have, and also doesn't add confusion on why we are calling `delete` on some seemingly arbitrary field. We can do something like:
```ts
const response = await resumeController.getAllVisibleResumes(admin);
expect(response).toStrictEqual({
  uuid: resume.uuid,
  user: member.uuid,
  isResumeVisible: resume.isResumeVisible
  ... // fill out for rest of fields
})
```

File: types/ApiResponses.ts
Comment (shravanhariharan2) [inline types/ApiResponses.ts:326]: nit: this would need to be `PublicResume[]` after we change our API to return public resumes

File: api/controllers/ResumeController.ts
Comment (shravanhariharan2) [inline api/controllers/ResumeController.ts:65]: nit: let's move the map operation to the return statement for easier readibility:
```suggestion
    const resumes = await this.resumeService.getVisibleResumes();
    return { error: null, resumes: resumes.map((resume) => resume.getPublicResume()) };
```

File: repositories/ResumeRepository.ts
Comment (shravanhariharan2) [inline repositories/ResumeRepository.ts:10]: Referring to the eager loading comment earlier, let's do the user join here

File: tests/resume.test.ts
Comment (shravanhariharan2) [inline tests/resume.test.ts:58]: nit: vocabulary
```suggestion

  test('admins cannot get hidden resumes', async () => {
```

File: tests/resume.test.ts
Comment (shravanhariharan2) [inline tests/resume.test.ts:80]: nit: formatting
```suggestion
    await new PortalState()
      .createUsers(member)
      .write();
```

File: types/ApiResponses.ts
Comment (shravanhariharan2) [inline types/ApiResponses.ts:1]: nit: this needs to be a relative URL instead of absolute, since our build system doesn't properly resolve absolute paths

File: types/ApiResponses.ts
Comment (shravanhariharan2) [inline types/ApiResponses.ts:307]: nit: this should also be `PublicResume[]`, since in almost every case we don't want to return a model object to the client since it might display fields that we didn't intend on showing (e.g. `id`, `UserModel`, etc). Even though this is a "private profile", we still return the "public" type because the term "public" in our types refers to anything the client interacts with

File: types/ApiResponses.ts
Comment (shravanhariharan2) [inline types/ApiResponses.ts:399]: nit: let's make this optional since not all `PublicResumes` need to return a `user` object (e.g. in the `getFullUserProfile()` case, resumes are queried by the user to populate `user.resumes`, so we don't need `user.resumes.user`.

**Review by shravanhariharan2 — COMMENTED** [review]
Mostly small comments, holding off on approval still since there's 1-2 breaking changes that need to be addressed, but should be good to go after changes are addressed

**Comment by dowhep** [comment]
Somehow merging the master branch updates make this branch fail test for "connection failed"

File: models/ResumeModel.ts
Comment (shravanhariharan2) [reply models/ResumeModel.ts:27]: We still want this field to be optional, since `this.user` might not always be defined. So e.g. when a user tries to view their own resumes, which is an operation where the Resume <-> User join is not done, then this line of code will throw an error. We need to do something like:

```ts
const publicResume: PublicResume = {
  uuid: this.uuid,
  isResumeVisible: this.isResumeVisible,
  url: this.url,
  lastUpdated: this.lastUpdated
}
if (this.user) publicResume.user = this.user.getPublicProfile();
return publicResume
```

And `PublicResume` should have `user` as an optional field. Let me know if there's any confusion around this.

File: models/UserModel.ts
Comment (shravanhariharan2) [reply models/UserModel.ts:138]: Apologies for revisiting such an old comment, but re-visiting this code I don't think what I recommended here is actually good practice. I think having a field here un-explicitly always returning `null` is something very easy to look over, so instead, I'd rather us just outright not specify the field here and instead make `resumes` optional in `PrivateProfile`. That way, we can return this object without the `resumes` field, and we can continue to add it as we do in `UserAccountService.getFullUserProfile`. I think this way gets rid of any ambiguity the code has by always returning `null`.

We can also remove the comment above mentioning that "resumes is null by design".

File: types/ApiResponses.ts
Comment (shravanhariharan2) [inline types/ApiResponses.ts:308]: As per above comment, let's make this optional

**Review by shravanhariharan2 — COMMENTED** [review]
Last few small code comments addressing optionality of fields. Otherwise LGTM, will approve after these are addressed

**Comment by shravanhariharan2** [comment]
> Somehow merging the master branch updates make this branch fail test for "connection failed"

Hmm I've never actually seen this test error before - seems like when I retried the run, the tests worked properly, but I will investigate what might've been happening here a bit more. I don't think it's anything related to this specific PR

**Review by shravanhariharan2 — APPROVED** [review]
lgtm

**Pull request merged**
